lib/services/saml.go

```diff
@@ -0,0 +1,191 @@
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package services
+
+import (
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/base64"
+	"encoding/xml"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/lib/tlsca"
+	"github.com/gravitational/teleport/lib/utils"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	saml2 "github.com/russellhaering/gosaml2"
+	"github.com/russellhaering/gosaml2/types"
+	dsig "github.com/russellhaering/goxmldsig"
+)
+
+// ValidateSAMLConnector validates the SAMLConnector and sets default values
+func ValidateSAMLConnector(sc SAMLConnector) error {
+	if err := sc.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+	if _, err := GetSAMLServiceProvider(sc, clockwork.NewRealClock()); err != nil {
```

Comment:
Is there a way to configure the clock? I've seen this method used in tests, using real clock might prevent time-based tests.